sls-versions/src/main/java/com/palantir/sls/versions/SlsVersionMatcherParser.java

```diff
@@ -0,0 +1,181 @@
+ * primitive longs live on the stack and don't impact GC.
+ */
+final class SlsVersionMatcherParser {
+
+    private static final int MAGIC_X_NUMBER = -1;
+
+    public static Optional<SlsVersionMatcher> safeValueOf(String string) {
+        OptionalInt major = OptionalInt.empty();
+        OptionalInt minor = OptionalInt.empty();
+        OptionalInt patch = OptionalInt.empty();
+
+        // major
+        long result = numberOrX(string, 0);
+        if (failed(result)) {
+            return Optional.empty(); // reject
+        }
+        if (getResult(result) != MAGIC_X_NUMBER) {
+            major = OptionalInt.of(getResult(result));
+        }
+
+        // dot
+        result = literalDot(string, getIndex(result));
+        if (failed(result)) {
+            return Optional.empty();
+        }
+
+        // minor
+        result = numberOrX(string, getIndex(result));
+        if (failed(result)) {
+            return Optional.empty(); // reject
+        }
+        if (getResult(result) != MAGIC_X_NUMBER) {
+            minor = OptionalInt.of(getResult(result));
+        }
+
+        // dot
+        result = literalDot(string, getIndex(result));
+        if (failed(result)) {
+            return Optional.empty();
+        }
+
+        // patch
+        result = numberOrX(string, getIndex(result));
+        if (failed(result)) {
+            return Optional.empty(); // reject
+        }
+        if (getResult(result) != MAGIC_X_NUMBER) {
+            patch = OptionalInt.of(getResult(result));
+        }
+
+        if (getIndex(result) < string.length()) {
+            return Optional.empty(); // reject due to trailing stuff
+        }
+
+        return SlsVersionMatcher.maybeCreate(string, major, minor, patch);
+    }
+
+    // "x" is signified by the magic negative number -1, which is distinct from Integer.MIN_VALUE which is a failure
+    private static long numberOrX(String string, int startIndex) {
+        long xResult = literalX(string, startIndex);
+        if (isOk(xResult)) {
+            return ok(getIndex(xResult), MAGIC_X_NUMBER);
+        }
+
+        long numberResult = number(string, startIndex);
+        if (isOk(numberResult)) {
+            return numberResult;
+        }
+
+        return fail(startIndex);
+    }
+
+    private static long number(String string, int startIndex) {
+        int next = startIndex;
+        int len = string.length();
+        while (next < len) {
+            int codepoint = string.codePointAt(next);
+            if (Character.isDigit(codepoint)) {
+                next += 1;
+            } else {
+                break;
+            }
+        }
+        if (next == startIndex) {
+            return fail(startIndex);
+        } else if (next == startIndex + 1) {
+            return ok(next, Character.digit(string.codePointAt(startIndex), 10));
+        } else {
+            try {
+                return ok(next, Integer.parseUnsignedInt(string.substring(startIndex, next)));
```

Comment:
for matches that might contain multiple digits (e.g. `10.x.x`), there is actually a new `Integer.parseUnsignedInt` method available from java9 onwards which would let me pass in this string and some indexes without needing to allocated substring.

In practise, I don't really care because the _vast_ majority of version matchers use single digit components.